src/components/tools/pen.js

```diff
@@ -185,68 +167,106 @@ AFRAME.registerComponent("pen", {
-
-      if (interaction.state.rightHand.held === this.el.parentNode) {
-        this.grabberId = "player-right-controller";
-      } else if (interaction.state.leftHand.held === this.el.parentNode) {
-        this.grabberId = "player-left-controller";
-      } else if (interaction.state.rightRemote.held === this.el.parentNode) {
-        this.grabberId = "cursor";
-        this.drawMode = DRAW_MODE.PROJECTION;
-      } else {
-        this.grabberId = null;
+      const isMine =
+        this.el.parentEl.components.networked.initialized && this.el.parentEl.components.networked.isMine();
+
+      if (this.penTip.material.visible !== isMine) {
+        this.penTip.material.visible = isMine;
       }
 
-      if (this.grabberId && pathsMap[this.grabberId]) {
-        const sfx = this.el.sceneEl.systems["hubs-systems"].soundEffectsSystem;
-        const paths = pathsMap[this.grabberId];
-        if (userinput.get(paths.startDrawing)) {
-          this._startDraw();
-          sfx.playSoundOneShot(SOUND_PEN_START_DRAW);
-        }
-        if (userinput.get(paths.stopDrawing)) {
-          this._endDraw();
-          sfx.playSoundOneShot(SOUND_PEN_STOP_DRAW);
-        }
-        const penScaleMod = userinput.get(paths.scalePenTip);
-        if (penScaleMod) {
-          this._changeRadius(penScaleMod);
-        }
-        if (userinput.get(paths.penNextColor)) {
-          this._changeColor(1);
-          sfx.playSoundOneShot(SOUND_PEN_CHANGE_COLOR);
-        }
-        if (userinput.get(paths.penPrevColor)) {
-          this._changeColor(-1);
-          sfx.playSoundOneShot(SOUND_PEN_CHANGE_COLOR);
-        }
-        if (userinput.get(paths.undoDrawing)) {
-          this._undoDraw();
-          sfx.playSoundOneShot(SOUND_PEN_UNDO_DRAW);
-        }
-        if (paths.switchDrawMode && userinput.get(paths.switchDrawMode)) {
-          this.drawMode = this.drawMode === DRAW_MODE.DEFAULT_2D ? DRAW_MODE.PROJECTION : DRAW_MODE.DEFAULT_2D;
+      if (isMine) {
+        this._handleInput();
+
+        const cursorPose =
+          this.data.drawMode === DRAW_MODE.PROJECTION && this.grabberId === "cursor"
+            ? AFRAME.scenes[0].systems.userinput.get(pathsMap.cursor.pose)
+            : null;
+
+        const intersection = this._getIntersection(cursorPose);
+
+        this._updatePenTip(intersection);
+
+        const remoteLaserVisible = this.data.drawMode === DRAW_MODE.PROJECTION && !!intersection;
+        const laserVisible = this.el.sceneEl.is("vr-mode") && !!intersection && remoteLaserVisible;
+        this.el.setAttribute("pen-laser", {
+          laserVisible: laserVisible,
+          remoteLaserVisible: remoteLaserVisible
+        });
+
+        if (remoteLaserVisible || laserVisible) {
+          this._updateLaser(cursorPose, intersection);
         }
+
+        const penVisible = this.grabberId !== "cursor" || !intersection;
+        this._setPenVisible(penVisible);
+        this.el.setAttribute("pen", { penVisible: penVisible });
```

Comment:
similarly here - check if this has changed to avoid calling `setAttribute` whenever possible. also consider using the `clobber` option in order to avoid creating a new object each time